src/sage/rings/polynomial/q_integer_valued_polynomials.py

```diff
@@ -0,0 +1,1255 @@
+        Univariate Polynomial Ring in q over Integer Ring
+
+    TESTS::
+
+        sage: from sage.rings.polynomial.q_integer_valued_polynomials import q_binomial_x
+        sage: q_binomial_x(4,2,1)
+        1/2*x^2 + 7/2*x + 6
+    """
+    if q is None:
+        ring_q = PolynomialRing(ZZ, 'q')
+    else:
+        ring_q = q.parent()
+    ring = PolynomialRing(ring_q.fraction_field(), 'x')
+    if n == 0:
+        return ring.one()
+    return ring.prod(q_int_x(m + 2 - i, q) / q_int(i, q)
+                     for i in range(1, n + 1))
+
+
+class QuantumValuedPolynomialRing(UniqueRepresentation, Parent):
+    r"""
+    The quantum-valued polynomial ring over a base ring.
+
+    Quantum-valued polynomial rings are commutative and associative
+    algebras, with a basis indexed by nonnegative integers.
+
+    The elements are polynomials in one variable `x` with coefficients in
+    the field of rational functions in `q`, such that the value at
+    every nonegative `q`-integer is a polynomial in `q`.
+
+    This algebra is endowed with two bases, named ``B`` or ``Binomial``
+    and ``S`` or ``Shifted``.
+
+    INPUT:
+
+    - ``R`` -- commutative ring
+
+    - ``q`` -- optional variable
+
+    There are two possible input formats:
+
+    - If the argument ``q`` is not given, then the ring ``R`` is
+      taken as a base ring and the ring of Laurent polynomials in `q`
+      over ``R`` is built and used.
+
+    - If the argument ``q`` is given, then it should belong to the ring ``R``
+      and be invertible in this ring.
+
+    EXAMPLES::
+
+        sage: F = QuantumValuedPolynomialRing(QQ).S(); F
+        Quantum-Valued Polynomial Ring over Rational Field
+        in the shifted basis
+
+        sage: F.gen()
+        S[1]
+
+        sage: S = QuantumValuedPolynomialRing(ZZ); S
+        Quantum-Valued Polynomial Ring over Integer Ring
+        sage: S.base_ring()
+        Univariate Laurent Polynomial Ring in q over Integer Ring
+
+    Quantum-valued polynomial rings commute with their base ring::
+
+        sage: K = QuantumValuedPolynomialRing(QQ).S()
+        sage: a = K.gen()
+        sage: c = K.monomial(2)
+
+    Quantum-valued polynomial rings are commutative::
+
+        sage: c^3 * a == c * a * c * c
+        True
+
+    We can also manipulate elements in the basis and coerce elements from our
+    base field::
+
+        sage: F = QuantumValuedPolynomialRing(QQ).S()
+        sage: B = F.basis()
+        sage: B[2] * B[3]
+        (q^-5+q^-4+q^-3)*S[3] - (q^-6+2*q^-5+3*q^-4+3*q^-3+2*q^-2+q^-1)*S[4]
+        + (q^-6+q^-5+2*q^-4+2*q^-3+2*q^-2+q^-1+1)*S[5]
+        sage: 1 - B[2] * B[2] / 2
+        S[0] - (1/2*q^-3)*S[2] + (1/2*q^-4+q^-3+q^-2+1/2*q^-1)*S[3]
+        - (1/2*q^-4+1/2*q^-3+q^-2+1/2*q^-1+1/2)*S[4]
+    """
+    @staticmethod
+    def __classcall_private__(cls, R, q=None) -> None:
+        """
+        Normalize the input.
+        """
+        if R not in Rings().Commutative():
```

Comment:
```suggestion
    def __classcall_private__(cls, R, q=None) -> None:
        """
        Normalize the input.

        EXAMPLES::

            sage: q = LaurentPolynomialRing(QQ, 'q').gen()
            sage: F1 = QuantumValuedPolynomialRing(QQ)
            sage: F2 = QuantumValuedPolynomialRing(q.parent(), q)
            sage: F1 is F2
            True
        """
        if R not in Rings().Commutative():
```
Thank you. Now just to add a doctest verifying the consistency. Once done, this will be a positive review.